File: logger.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
# Copyright 2016 Ryan Sawhill Aroha rsaw@redhat.com
# See https://github.com/ryran/b19scripts/ for latest version
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Modules from standard library
from __future__ import print_function
import argparse, logging, logging.handlers
from socket import SOCK_DGRAM, SOCK_STREAM
from sys import exit, stdin, stderr
from os import getuid
from pwd import getpwuid

# Set some globals
prog = 'logger.py'
versionNumber = '0.2.0'
versionDate = '2016/04/11'
description = "A simplistic python logger clone that reads input via stdin"
epilog = "{} v{} last mod {};".format(prog, versionNumber, versionDate)
epilog += "\nFor issues & questions, see: https://github.com/ryran/b19scripts/issues"
user = getpwuid(getuid()).pw_name

# Add missing syslog levels to python logger
logging.addLevelName(25, 'NOTICE')
logging.addLevelName(60, 'ALERT')
logging.addLevelName(70, 'EMERG')

# Customize logging.handlers.SysLogHandler.priority_map to include added levels
loggerSyslogHandlerPriorityMap = {
    'DEBUG': 'debug',
    'INFO': 'info',
    'NOTICE': 'notice',
    'WARNING': 'warning',
    'ERROR': 'error',
    'CRITICAL': 'critical',
    'ALERT': 'alert',
    'EMERG': 'emerg',
    }

# Build a dict to establish what integer to pass to Logger.log()
syslogPriorityToLoggerLevel = {
    'debug': 10,
    'info': 20,
    'notice': 25,
    'warning': 30,
    'warn': 30,
    'error': 40,
    'err': 40,
    'critical': 50,
    'crit': 50,
    'alert': 60,
    'emerg': 70,
    'panic': 70,
    }

class CustomFormatter(argparse.ArgumentDefaultsHelpFormatter):
    """This custom formatter eliminates the duplicate metavar in help lines."""
    def _format_action_invocation(self, action):
        if not action.option_strings:
            metavar, = self._metavar_formatter(action, action.dest)(1)
            return metavar
        else:
            parts = []
            if action.nargs == 0:
                parts.extend(action.option_strings)
            else:
                default = action.dest.upper()
                args_string = self._format_args(action, default)
                for option_string in action.option_strings:
                    parts.append('%s' % option_string)
                parts[-1] += ' %s'%args_string
            return ', '.join(parts)

def parse_cmdline():
    fmt = lambda prog: CustomFormatter(prog)
    p = argparse.ArgumentParser(
        prog=prog,
        description=description,
        add_help=True,
        epilog=epilog,
        formatter_class=fmt)
    p.add_argument('-p', '--priority', metavar='PRIO', dest='priostring', default='user.info', help="specify priority for given message")
    p.add_argument('-t', '--tag', default=user, help="add tag to given message")
    p.add_argument('-i', '--id', action='store_true', help="add process ID to tag")
    g1 = p.add_mutually_exclusive_group()
    g1.add_argument('-u', '--socket', default='/dev/log', help="write to local UNIX socket")
    g1.add_argument('-d', '--udp', action='store_true', help="log via UDP instead of UNIX socket")
    g1.add_argument('-T', '--tcp', action='store_true', help="log via TCP instead of UNIX socket")
    p.add_argument('-n', '--server', default='localhost', help="DNS/IP of syslog server to use with --udp or --tcp")
    p.add_argument('-P', '--port', type=int, default=514, help="port to use with --udp or --tcp")
    opts =  p.parse_args()
    try:
        opts.facility = opts.priostring.split('.')[0]
        logging.handlers.SysLogHandler.facility_names[opts.facility]
        opts.priority = opts.priostring.split('.')[1]
        logging.handlers.SysLogHandler.priority_names[opts.priority]
        opts.level = syslogPriorityToLoggerLevel[opts.priority]
    except:
        print("Improper 'priority' specified")
        print("Must be <facility>.<priority> as described in logger(1) man page\n")
        raise
    return opts

def main():
    opts = parse_cmdline()
    myLogger = logging.getLogger(opts.tag)
    myLogger.setLevel(logging.DEBUG)
    if opts.udp:
        myHandler = logging.handlers.SysLogHandler(address=(opts.server, opts.port), facility=opts.facility, socktype=SOCK_DGRAM)
    elif opts.tcp:
        myHandler = logging.handlers.SysLogHandler(address=(opts.server, opts.port), facility=opts.facility, socktype=SOCK_STREAM)
    else:
        myHandler = logging.handlers.SysLogHandler(address=opts.socket, facility=opts.facility)
    myHandler.priority_map = loggerSyslogHandlerPriorityMap
    if opts.id:
        myFormatter = logging.Formatter('%(name)s[%(process)s]: %(message)s')
    else:
        myFormatter = logging.Formatter('%(name)s: %(message)s')
    myHandler.setFormatter(myFormatter)
    myLogger.addHandler(myHandler)
    while 1:
        line = stdin.readline()
        if not line:
            break
        myLogger.log(opts.level, line)

if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print("\nReceived KeyboardInterrupt. Exiting.")
        exit()

File: logger.pl
#!/usr/bin/perl
#
# Copyright 2016 Ryan Sawhill Aroha rsaw@redhat.com
# See https://github.com/ryran/b19scripts/ for latest version
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

use strict;
use warnings;
use Getopt::Long qw( GetOptions );
use Sys::Syslog qw( :DEFAULT setlogsock );

Getopt::Long::Configure ("gnu_getopt");

my $print_help;
my $tag = (getpwuid($<))[0];
my $prio_string = 'user.info';
my $facility = '';
my $priority = '';
my $socket = '/dev/log';
my $use_stderr;
my $use_pid;
my $use_tcp;
my $use_udp;
my $syslog_server= 'localhost';
my $port = 514;
my $openlog_opts = 'cons';
my $log;
my @facilities = qw( auth authpriv cron daemon ftp lpr mail news syslog user uucp local0 local1 local2 local3 local4 local5 local6 local7 );
my @priorities = qw( debug info notice warning warn err crit alert emerg panic );

sub get_usage {
    (my $usage = <<"    EOF") =~ s/^ {8}//gm;
        usage: logger.pl [-h|--help] [-p PRIO] [-t TAG] [-i] [-u SOCKET | --udp | --tcp]
                         [-n SERVER] [-P PORT]
    EOF
    return $usage;
}

sub get_help {
    (my $help = <<"    EOF") =~ s/^ {8}//gm;
        
        A simplistic perl logger clone that reads input via stdin
        
        optional arguments:
          -h, --help           show this help message and exit
          -p, --priority PRIO  specify priority for given message (default: $prio_string)
          -t, --tag TAG        add tag to given message (default: $tag)
          -i, --id             add process ID to tag (default: False)
          -u, --socket SOCKET  write to local UNIX socket (default: $socket)
          -d, --udp            log via UDP instead of UNIX socket (default: False)
          -T, --tcp            log via TCP instead of UNIX socket (default: False)
          -n, --server SERVER  DNS/IP of syslog server to use with --udp or --tcp
                               (default: $syslog_server)
          -P, --port PORT      port to use with --udp or --tcp (default: $port)
          -s, --stderr         output to standard error as well (default: False)
        
        logger.pl v0.1.0 last mod 2016/04/11
        For issues & questions, see: https://github.com/ryran/b19scripts/issues
    EOF
    return $help;
}

sub print_prio_err {
    print "Improper 'priority' specified\n";
    print "Must be <facility>.<priority> as described in logger(1) man page\n";
}

# Parse options
GetOptions(
    'help|h' => \$print_help,
    'tcp|T' => \$use_tcp,
    'udp|d' => \$use_udp,
    'id|i' => \$use_pid,
    'server|n=s' => \$syslog_server,
    'port|P=i' => \$port,
    'priority|p=s' => \$prio_string,
    'stderr|s' => \$use_stderr,
    'tag|t=s' => \$tag,
    'socket|u=s' => \$socket,
) or die get_usage();

# Check arguments
if (length $print_help) {
    print get_usage();
    print get_help();
    exit;
}
($facility, $priority) = split('\.', $prio_string);
if ($facility eq '' || $priority eq '') {
    print_prio_err();
    exit 1;
}
my %facilities_set;
@facilities_set{@facilities} = ();
my %priorities_set;
@priorities_set{@priorities} = ();
if (exists $facilities_set{$facility} && exists $priorities_set{$priority}) {
} else {
    print_prio_err();
    exit 1;
}
if (length $use_tcp && length $use_udp) {
    print "Can't use --tcp and --udp at the same time\n";
    exit 1;
}

# Setup log socket
if ($socket eq '/dev/log') {
} elsif (length $use_tcp) {
    setlogsock({ type => 'tcp', port => $port, host => $syslog_server });
} elsif (length $use_udp) {
    setlogsock({ type => 'udp', port => $port, host => $syslog_server });
} else {
    setlogsock({ type => 'unix', path => $socket });
}

# Setup openlog opts
if (length $use_pid) {
    $openlog_opts .= ',pid';
}
if (length $use_stderr) {
    $openlog_opts .= ',perror';
}

# Open connection and start reading stdin
openlog($tag, $openlog_opts, $facility);
while ($log = <STDIN>) {
    syslog($priority, $log);
}
closelog;
